Custom error message when gradle-plugin is applied to an Android project

I've just started to play around with the new alpha and just bumped the version number which yielded the following exception:

> Caused by: org.gradle.api.UnknownDomainObjectException: SourceSet with name 'main' not found.

The underlying cause is https://github.com/square/sqldelight/blob/7d824b741a001c47a44359dc44997bee820efdba/sqldelight-gradle/plugin/src/main/kotlin/com/squareup/sqldelight/gradle/SqlDelightPlugin.kt#L36

Should we throw a custom exception saying that this plugin won't work with android and hence the `android-gradle-plugin`  has to be used?

At least once 1.0.0 is ready it should be put into a migration guide.

--- a/sqldelight-gradle/android-plugin/src/main/kotlin/com/squareup/sqldelight/gradle/SqlDelightAndroidPlugin.kt
+++ b/sqldelight-gradle/android-plugin/src/main/kotlin/com/squareup/sqldelight/gradle/SqlDelightAndroidPlugin.kt
@@ -33,6 +33,22 @@ import java.io.File
 class SqlDelightAndroidPlugin : SqlDelightPlugin() {
   override fun apply(project: Project) {
     project.extensions.create("sqldelight", SqlDelightExtension::class.java)
+    if (!project.plugins.hasPlugin("android")) {
+      throw IllegalStateException(
+          """
+      Kotlin projects need to apply the sqldelight kotlin plugin:
+
+      buildscript {
+        dependencies {
+          classpath "com.squareup.sqldelight:gradle-plugin:$VERSION
+        }
+      }
+
+      apply plugin: "com.squareup.sqldelight"
+      """.trimIndent()
+      )
+    }
+
     project.plugins.all {
       when (it) {
         is AppPlugin -> {
@@ -52,8 +68,6 @@ class SqlDelightAndroidPlugin : SqlDelightPlugin() {
     if (System.getProperty("sqldelight.skip.runtime") != "true") {
       compileDeps.add(project.dependencies.create("com.squareup.sqldelight:runtime-jdk:$VERSION"))
     }
-    compileDeps.add(
-        project.dependencies.create("com.android.support:support-annotations:23.1.1"))
 
     var packageName: String? = null
     val sourceSets = mutableListOf<List<String>>()

--- a/sqldelight-gradle/android-plugin/src/test/kotlin-project/build.gradle
+++ b/sqldelight-gradle/android-plugin/src/test/kotlin-project/build.gradle
@@ -1,0 +1,4 @@
+plugins {
+  id 'kotlin'
+  id 'com.squareup.sqldelight.android'
+}

--- a/sqldelight-gradle/android-plugin/src/test/kotlin/com/squareup/sqldelight/PluginTest.kt
+++ b/sqldelight-gradle/android-plugin/src/test/kotlin/com/squareup/sqldelight/PluginTest.kt
@@ -1,0 +1,32 @@
+package com.squareup.sqldelight
+
+import com.google.common.truth.Truth
+import org.gradle.testkit.runner.GradleRunner
+import org.junit.Test
+import java.io.File
+
+class PluginTest {
+  @Test
+  fun `Applying the android plugin without AGP applied throws`() {
+    val fixtureRoot = File("src/test/kotlin-project")
+    val runner = GradleRunner.create()
+        .withProjectDir(fixtureRoot)
+        .withPluginClasspath()
+
+    val result = runner
+        .withArguments("build", "--stacktrace", "-Dsqldelight.skip.runtime=true")
+        .buildAndFail()
+    Truth.assertThat(result.output)
+        .contains("""
+      Kotlin projects need to apply the sqldelight kotlin plugin:
+
+      buildscript {
+        dependencies {
+          classpath "com.squareup.sqldelight:gradle-plugin:$VERSION
+        }
+      }
+
+      apply plugin: "com.squareup.sqldelight"
+      """.trimIndent())
+  }
+}

--- a/sqldelight-gradle/plugin/build.gradle
+++ b/sqldelight-gradle/plugin/build.gradle
@@ -34,6 +34,7 @@ dependencies {
   bundled deps.intellijCore
 
   fixtureClasspath deps.plugins.kotlin
+  fixtureClasspath deps.plugins.android
 }
 
 jar {

--- a/sqldelight-gradle/plugin/src/main/kotlin/com/squareup/sqldelight/gradle/SqlDelightPlugin.kt
+++ b/sqldelight-gradle/plugin/src/main/kotlin/com/squareup/sqldelight/gradle/SqlDelightPlugin.kt
@@ -15,6 +15,7 @@
  */
 package com.squareup.sqldelight.gradle
 
+import com.squareup.sqldelight.VERSION
 import com.squareup.sqldelight.core.SqlDelightPropertiesFile
 import com.squareup.sqldelight.core.lang.MigrationFileType
 import com.squareup.sqldelight.core.lang.SqlDelightFileType
@@ -29,6 +30,22 @@ import java.io.File
 
 open class SqlDelightPlugin : Plugin<Project> {
   override fun apply(project: Project) {
+    if (project.plugins.hasPlugin("android")) {
+      throw IllegalStateException(
+          """
+      Android projects need to apply the sqldelight android plugin:
+
+      buildscript {
+        dependencies {
+          classpath "com.squareup.sqldelight:android-gradle-plugin:$VERSION
+        }
+      }
+
+      apply plugin: "com.squareup.sqldelight.android"
+      """.trimIndent()
+      )
+    }
+
     val extension = project.extensions.create("sqldelight", SqlDelightExtension::class.java)
     val outputDirectory = File(project.buildDir, "sqldelight")
 
@@ -92,6 +109,7 @@ open class SqlDelightPlugin : Plugin<Project> {
           it.outputDirectory = schemaOutputDirectory
           it.source(sourceSet)
           it.include("**${File.separatorChar}*.${SqlDelightFileType.defaultExtension}")
+          it.include("**${File.separatorChar}*.${MigrationFileType.defaultExtension}")
         }
       generateSchemaTask.group = "sqldelight"
       generateSchemaTask.description = "Generate a .db file containing the current schema."

--- a/sqldelight-gradle/plugin/src/test/android-project/build.gradle
+++ b/sqldelight-gradle/plugin/src/test/android-project/build.gradle
@@ -1,0 +1,22 @@
+plugins {
+  id 'com.android.application'
+  id 'com.squareup.sqldelight'
+}
+
+apply from: '../../../../../gradle/dependencies.gradle'
+
+android {
+  compileSdkVersion versions.compileSdk
+
+  sourceSets {
+    main.java.srcDirs += '../../../../../runtime/sqldelight-runtime/src/main/java'
+  }
+
+  lintOptions {
+    textReport true
+  }
+}
+
+sqldelight {
+  schemaOutputDirectory = file('src/main/sqldelight/databases')
+}

--- a/sqldelight-gradle/plugin/src/test/kotlin/com/squareup/sqldelight/PluginTest.kt
+++ b/sqldelight-gradle/plugin/src/test/kotlin/com/squareup/sqldelight/PluginTest.kt
@@ -1,0 +1,32 @@
+package com.squareup.sqldelight
+
+import com.google.common.truth.Truth
+import org.gradle.testkit.runner.GradleRunner
+import org.junit.Test
+import java.io.File
+
+class PluginTest {
+  @Test
+  fun `Applying the android plugin without AGP applied throws`() {
+    val fixtureRoot = File("src/test/android-project")
+    val runner = GradleRunner.create()
+        .withProjectDir(fixtureRoot)
+        .withPluginClasspath()
+
+    val result = runner
+        .withArguments("build", "--stacktrace", "-Dsqldelight.skip.runtime=true")
+        .buildAndFail()
+    Truth.assertThat(result.output)
+        .contains("""
+      Android projects need to apply the sqldelight android plugin:
+
+      buildscript {
+        dependencies {
+          classpath "com.squareup.sqldelight:android-gradle-plugin:$VERSION
+        }
+      }
+
+      apply plugin: "com.squareup.sqldelight.android"
+      """.trimIndent())
+  }
+}

